pydantic/exceptions.py

```diff
@@ -1,37 +1,58 @@
-import inspect
 import json
+from typing import Iterable, Type
 
 from .utils import to_snake_case
 
 __all__ = (
     'Error',
     'ValidationError',
     'ConfigError',
-    'Missing',
-    'Extra',
+
+    'ValueError_',
+    'TypeError_',
+
+    'MissingError',
+    'ExtraError',
+
+    'DecimalError',
+    'DecimalIsNotFiniteError',
+    'DecimalMaxDigitsError',
+    'DecimalMaxPlacesError',
+    'DecimalWholeDigitsError',
+
+    'UUIDError',
+    'UUIDVersionError',
 )
 
 
 class Error:
-    __slots__ = 'exc_info', 'loc'
+    __slots__ = 'exc', 'loc'
```

Comment:
Perhaps we can get rid of `Error` altogether and store `loc` on the Exception?

Reply:
I don't think that we need to get rid of `Error` also I get your idea with `_PydanticMixin` which contains all about error (`ctx`, `msg`, `loc`, ...), but I don't think that this is a good idea.

To my mind this is not a good idea because:
* `Error` and `ValidationError` are wrappers around pydantic's validation logic, they hides inside errors representation and things like `loc`;
* moving logic into `_PydanticMixin` will broke your idea with `TypeError` and `ValueError` which can be used by users to raise errors in user's validators;
* in case if we want to store `loc` inside error as you proposed we need to rewrite a huge part because for now exceptions know nothing about where they raised.

I think we need to keep `Error` and `ValidationError`, but we need to change some things, maybe just naming. Also I want to keep `Error` because it's a good wrapper/container to hide validation logic from users and it's also helps us to change anything inside without introducing breaking changes.

I propose:
* move all errors like `MissingError`, `ExtraError`, `DecimalError` to `errors.py` module;
* rename `ValidationError` to `ValidationErrors`, because it's a container of errors;
* rename `Error` to `ValidationError`, because  it's a error container which knows all about error (exception and location) and it's a good place to work with config to get custom message templates;
* rename `exceptions.py` to something more descriptive? like `error_wrappers.py`?

Reply:
> moving logic into `_PydanticMixin` will break your idea with `TypeError` and `ValueError` which can be used by users to raise errors in user's validators

Very good point, let's Keep `Error`

> * move all errors like `MissingError`, `ExtraError`, `DecimalError` to `errors.py` module;

Okay

* rename `Error` to `FieldError`, because  it's a error container which knows all about field error (exception and location) and it's a good place to work with config to get custom message templates;

I think `Error` is fine.

* rename `exceptions.py` to something more descriptive? like `error_wrappers.py`?

Okay